exporter/signalfxexporter/translation/converter.go

```diff
@@ -94,287 +83,273 @@ func NewMetricsConverter(logger *zap.Logger, t *MetricTranslator) *MetricsConver
+		dps = convertIntDatapoints(metric.IntGauge().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeIntSum:
+		dps = convertIntDatapoints(metric.IntSum().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeDoubleGauge:
+		dps = convertDoubleDatapoints(metric.DoubleGauge().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeDoubleSum:
+		dps = convertDoubleDatapoints(metric.DoubleSum().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeIntHistogram:
+		dps = convertIntHistogram(metric.IntHistogram().DataPoints(), basePoint, extraDimensions)
+	case pdata.MetricDataTypeDoubleHistogram:
+		dps = convertDoubleHistogram(metric.DoubleHistogram().DataPoints(), basePoint, extraDimensions)
+	}
 
-		for _, series := range metric.Timeseries {
-			dimensions := make([]*sfxpb.Dimension, numLabels+len(extraDimensions))
-			copy(dimensions, extraDimensions)
-			for i := 0; i < numLabels; i++ {
-				dimension := &sfxpb.Dimension{
-					Key:   descriptor.LabelKeys[i].Key,
-					Value: series.LabelValues[i].Value,
-				}
-				dimensions[len(extraDimensions)+i] = dimension
-			}
+	if c.metricTranslator != nil {
+		dps = c.metricTranslator.TranslateDataPoints(c.logger, dps)
+	}
 
-			for _, dp := range series.Points {
-
-				var msec int64
-				if dp.Timestamp != nil {
-					msec = dp.Timestamp.Seconds*1e3 + int64(dp.Timestamp.Nanos)/1e6
-				}
-
-				sfxDataPoint := &sfxpb.DataPoint{
-					// Source field is not set by code seen at
-					// github.com/signalfx/golib/sfxclient/httpsink.go
-					Metric:     metricName,
-					MetricType: metricType,
-					Timestamp:  msec,
-					Dimensions: dimensions,
-				}
-
-				switch pv := dp.Value.(type) {
-				case *metricspb.Point_Int64Value:
-					sfxDataPoint.Value = sfxpb.Datum{IntValue: &pv.Int64Value}
-					metricDataPoints = append(metricDataPoints, sfxDataPoint)
-
-				case *metricspb.Point_DoubleValue:
-					sfxDataPoint.Value = sfxpb.Datum{DoubleValue: &pv.DoubleValue}
-					metricDataPoints = append(metricDataPoints, sfxDataPoint)
-
-				case *metricspb.Point_DistributionValue:
-					metricDataPoints = appendDistributionValues(
-						metricDataPoints,
-						sfxDataPoint,
-						pv.DistributionValue)
-				case *metricspb.Point_SummaryValue:
-					metricDataPoints = appendSummaryValues(
-						metricDataPoints,
-						sfxDataPoint,
-						pv.SummaryValue)
-				default:
-					numDroppedTimeSeries++
-					c.logger.Warn(
-						"Timeseries dropped to unexpected metric type",
-						zap.String("metric", sfxDataPoint.Metric))
-				}
+	return dps, 0
+}
 
-			}
-		}
+func labelsToDimensions(labels pdata.StringMap, extraDims []*sfxpb.Dimension) []*sfxpb.Dimension {
+	dimensions := make([]*sfxpb.Dimension, len(extraDims), labels.Len()+len(extraDims))
+	copy(dimensions, extraDims)
+	labels.ForEach(func(k string, v pdata.StringValue) {
+		dimensions = append(dimensions, &sfxpb.Dimension{
+			Key:   k,
+			Value: v.Value(),
+		})
+	})
+
```

Comment:
Consider to avoid individual allocations, and unnecessary copy:

```golang
	rawDimensions := make([]*sfxpb.Dimension, labels.Len())
	i := 0
	labels.ForEach(func(k string, v pdata.StringValue) {
		rawDimensions[i].Key = k
		rawDimensions[i].Value = v.Value()
		dimensions = append(dimensions, &rawDimensions[i])
	})
```

Reply:
See https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1375#discussion_r513438802

Reply:
Not sure it is related, this is an optimization for the labels not for the extra labels.